trio/_file_io.py

```diff
@@ -48,7 +74,132 @@
 }
 
 
-class AsyncIOWrapper(AsyncResource):
+FileT = TypeVar("FileT")
+FileT_co = TypeVar("FileT_co", covariant=True)
+T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
+T_contra = TypeVar("T_contra", contravariant=True)
+AnyStr_co = TypeVar("AnyStr_co", str, bytes, covariant=True)
+AnyStr_contra = TypeVar("AnyStr_contra", str, bytes, contravariant=True)
+
+# Define a protocol for every method/property we expose. Each is effectively a predicate checking
+# whether a class defines this method/property.
```

Comment:
This .. is kinda magic to me, and probably to lots of other people as well. Could you add a couple more lines explaining how/why this works?